### PR TITLE
Allow dnsmasq caching config to be modified

### PR DESCRIPTION
**Pull request by ventifus**
### Which issue this PR addresses:

Fixes https://issues.redhat.com/browse/ARO-2427. Introduces a new field to ClusterSpec `dnsmasq` that allows configuring of dnsmasq cache-size, min-cache-ttl, max-cache-ttl and no-negcache.

Depends on 
- release-4.10-azure: https://github.com/jewzaam/installer-aro/pull/20
- release-4.11-azure: https://github.com/jewzaam/installer-aro/pull/19

### What this PR does / why we need it:

If a query is forwarded by dnsmasq is lost, then dnsmasq will hold on to that query for 40 seconds waiting for a response. If additional queries for that same name (and type) come in during that time, they will block on waiting for the first forwarded query. To reduce the number of upstream requests needed, this PR allows enabling and tuning caching.

### Test plan for issue:

Deploy to a test cluster. Let it run for a while (1 day or more) and check CoreDNS logs for "i/o timeout". **PENDING**

Install the new aro operator and make sure it doesn't update any machineconfig :heavy_check_mark: **PASS**

Test that each setting updates the machineconfig appropriately. Also verify the resulting dnsmasq.conf config file is formatted coorectly and dnsmasq loads.
  - cacheSize  :heavy_check_mark: **PASS**
  - maxCacheTTL  :heavy_check_mark: **PASS**
  - minCacheTTL :heavy_check_mark: **PASS**
  - noNegCache  :heavy_check_mark: **PASS**

### Is there any documentation that needs to be updated for this PR?
Internal documentation needs to be created detailing how to configure dnsmasq caching, when to enable it, and how to tune it for workload.
Once we have this deployed in the field solving real customer issues, we will consider enabling cache by default. Data will need to be collected to determine how big a cache to configure; the dnsmasq upstream default is 150 entries.

### CODE REVIEW

**Comment by dem4gus** [comment]
Were the changes in the vendor folder made upstream and are being incorporated here or are we just changing the vendor files directly?

**Comment by ventifus** [comment]
@dem4gus I'm changing the vendor files directly only to make it easier to test and review. I'll split out those changes to aro-installer once I'm sure this approach will work and I'm ready to take this out of draft.

File: go.mod
Comment (s-amann) [inline go.mod:459]: should this get reverted? or are you the new owner of `installer-aro`?

File: pkg/operator/deploy/deploy.go
Comment (ventifus) [reply pkg/operator/deploy/deploy.go]: Added

File: go.mod
Comment (ventifus) [reply go.mod:459]: The intent here was to make CI and such pass. This won't be needed and I'll revert once those changes are merged into installer-aro.

File: pkg/operator/deploy/deploy.go
Comment (ellis-johnson) [inline pkg/operator/deploy/deploy.go:223]: Are these just being set as 0 as a default value?

Comment (SudoBrendan) [reply pkg/operator/deploy/deploy.go:223]: +1 here, I feel like these should be interpreted from some kind of higher-level config, that we can modify without code changes, right? (envvars, RP-Config, etc)

**Review by SudoBrendan — CHANGES_REQUESTED** [review]
A great start! I still have open questions about caching in general and if that fits into our ARO strategy moving into the upstream installer, I'll let Ben comment, since I'm a little removed from the intricacies. If that aligns, the only change I'd like to see is some kind of configuration of DNSMasq that we can modify/rollout/revert with a config release rather than a code change.

**Comment by bennerv** [comment]
> I'd be curious to know if @bennerv sees any issues with us adding caching to dns as it relates to installer changes? I know that in a recent openshift/installer sync, we had mentioned that DNSMasq is a non-caching DNS server (ref: https://docs.google.com/presentation/d/1axmIrS3dyzAyGlLcJOly5vPFv1uymvu-jOy2uATYi_E/edit#slide=id.g136f68161c2_0_118), would this implementation change our ability to use upstream in the future?

Didn't get a chance to address this yet, but it's been in the back of my mind.  

Here's my issue with the caching; CoreDNS already leverages caching.  So by adding caching at the top-level (dnsmasq), we're introducing a chain of caches which can result in unexpected behavior from a DNS side.  If a DNS record is updated, the TTL on the DNS record isn't honored from Corefile, and then there's also caching at dnsmasq level.  

The reason dnsmasq was never meant to cache things is that we wanted the normal behavior to happen with the addition of dnsmasq.  Unfortunately it has had some unexpected consequences.  I would rather use the fast retries option if possible, but it seems that's not the case quite yet, and won't be unless we ask the RHEL team to backport.  I don't see this as an option short-term.  

I'm not sure if it's possible to force dnsmasq to use TCP to make it more reliable when querying upstream DNS servers.  That might be one possibility, otherwise we're stuck with this without using a different resolver or waiting for fast retries.




```
$ oc -n openshift-dns get cm dns-default  -o yaml
apiVersion: v1
data:
  Corefile: |
    .:5353 {
        bufsize 512
        cache 900 {
            denial 9984 30
        }
```

File: vendor/modules.txt
Comment (jhoreman) [inline vendor/modules.txt:881]: Assuming these should also be reverted as part of the above comment chain.

**Pull request closed without merging**
